add functions to change timeout

UziTech opened the pull request:
add functions to `req` to change timeout

`req.resetTimeout(delay)`: reset the timeout and start from now
`req.addTimeout(delay)`: add to the current timeout
`req.getTimeout()`: get the amount of milliseconds left (0 after timedout)

fixes #26

dougwilson commented:
Hi @UziTech thanks for the pull request! I'll start reviewing over the coming days. The following jumped out at me just with an initial glance at the pull request, though:

1. There is no documentation. Some folks wait to write documentation after the review, but you didn't mention that in your description, so wasn't sure if you were waiting or accidentally forgot to write it.
2. `req.setTimeout` already exists -- https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_message_settimeout_msecs_callback . You'll need to use a different method name as not to overwrite core Node.js method and break existing apps (and just otherwise prevent apps from being able to access this functionality).
3. It doesn't seem like `req.clearTimeout()` will stop `req.timeoutLeft()` from continuing to count down as if the timeout was still there.

UziTech commented:
Good points.

  1. I figured a few things would change so I was waiting on the docs.
  2. What would you like it to be called? A couple suggestions:
    a. `.startTimeout()`
    b. `.restartTimeout()`
    c. `.timeout()`
  3. Nice catch, I will fix that.

UziTech commented:
also a thought: should `.timeoutLeft()` be changed to `.timeoutRemaining()`

P.S. It's a good thing this problem doesn't involve cache invalidation.

UziTech commented:
I renamed the functions:

```
req.setTimeout(...) -> req.resetTimeout(...)
req.timeoutLeft(...) -> req.getTimeout(...)
```

and added docs to the README

UziTech commented:
@dougwilson any chance you got to review this?

UziTech commented:
🙃

UziTech commented:
Is anyone looking at this?

/cc @dougwilson @juliangruber @jonathanong

juliangruber commented:
not at this moment, sorry

UziTech commented:
You can remove the `need docs` label. I added docs to explain the new methods.

UziTech commented:
It would be great to get this merged.

/cc @dougwilson @juliangruber @jonathanong

dougwilson commented on index.js:
```diff
@@ -44,19 +44,52 @@ function timeout (time, options) {
       clearTimeout(id)
     }
 
+    req.resetTimeout = function (newDelay) {
```
If calling reset after the timeout already occurred, should req.timedout be reset back to false? Or should is stay true and update docs to describe it just represents the first trigger? Or not allow a reset after timeout happened?

UziTech replied:
it should probably remain false to let the user know that it was triggered.

UziTech replied:
It remains `true` if `req.clearTimeout()` is called after timeout is fired.

I updated the docs to reflect this.

dougwilson commented:
Sorry. I removed the label and left one comment about the reset behavior.

dougwilson commented on test/test.js:
```diff
@@ -106,6 +106,162 @@ describe('timeout()', function () {
+    })
+    it('should reset the timeout with 5000 default', function (done) {
+      var server = createServer(null,
+        function (req, res) { req.resetTimeout() },
```
How does this test validate that 5000 is what the reset used? It is identical to the first test just missing the argument.

It seems like if the impl was change to not accept an argument, all three of these tests still pass?

UziTech replied:
Good call. I updated the tests to test the actual timeout.

UziTech commented:
@dougwilson this should be good to go.

UziTech commented:
@dougwilson can this get reviewed?

UziTech commented:
It would be great to get this merged.

/cc @dougwilson @juliangruber @jonathanong

dougwilson commented:
Hi @UziTech sorry. There are still a few issues with the PR and I was just going to resolve them for you instead of doing more back and forth, I just haven't gotten to it yet.

UziTech commented:
@dougwilson can we get this merged?

UziTech commented:
@dougwilson are you still able to merge this?

UziTech commented:
Over a year and a half waiting for this PR to be merged!

@expressjs If you need help maintaining some of the expressjs packages I would be willing to help.

UziTech commented:
We are almost at the 2 year mark. Is there anything I can do to move this along?

dougwilson commented:
Hi @UziTech , I think that with a new sweeping overview we are doing, this module is likely to just be discontinued and deprecated, as it will not work at all with the upcoming Express, as it calls `next()` more than once and other design issues. I think we should just be real at this point: if this is something you need / want, just publish the module as a fork and go from there; even the name of this module is problematic for our organization, having the `connect-` prefix.

UziTech commented:
Thanks for the response.

dougwilson commented:
And @UziTech I hope you can accept my apology for letting this sit here -- I really didn't want to continue this module in many ways and the original authors didn't seem to be around responsive, either. If it does get a fresh breath of air, having the module not start with `connect-` would be awesome anyhow, as it makes little sense with Express since it hasn't been built on connect in years... I do have a couple legacy apps that use this module still, but there are design issues that makes it hard to use; the readme shows putting a "stopper" middleware between everything to at least make it _kind of_ work, but it... sucks :(

It may make sense to rebirth as like a middleware _wrapper_ instead. To do something like `app.use(timeout(200, myMiddleware))` that would time out that middleware at 200ms, IDK. I haven't put too much thought into it, and I think that is part of the problem--it deserves to be in the hands of someone who can really put thought into it...

UziTech commented:
No problem. I just want to make clear that, whatever timeout method is used, it is important to have the ability to cancel/extend the timeout in certain situations.

In my project a middleware wrapper wouldn't work because there are multiple middleware that could cause a timeout, so it works better to set a timeout over the entire request.

As I said above, If you need any help maintaining small projects I would be happy to help 😁 :+1:
